File: backend authentication/src/controllers/user.controller.ts
import { Request, Response } from "express";
import bcrypt from "bcryptjs";
import { supabase } from "../config/supabase";

// ─── User ID Generator ────────────────────────────────────────────────────────

async function generateUserId(
  role: string,
  school?: string,
  classGrade?: string
): Promise<string> {
  let prefix: string;

  switch (role) {
    case "subadmin":
      prefix = "SA";
      break;

    case "teacher":
      prefix = "TA";
      break;

    case "principal": {
      const letter = (school || "X").trim().charAt(0).toUpperCase();
      prefix = `P${letter}`;
      break;
    }

    case "student": {
      const letter = (school || "X").trim().charAt(0).toUpperCase();
      // e.g. "Class 9-A" → "9A", "9A" → "9A"
      const cleanClass = (classGrade || "")
        .replace(/class\s*/i, "")
        .replace(/[\s\-]/g, "")
        .toUpperCase();
      prefix = `S${letter}${cleanClass}`;
      break;
    }

    default:
      prefix = "USR";
  }

  const { count } = await supabase
    .from("user_auth")
    .select("*", { count: "exact", head: true })
    .like("user_id", `${prefix}%`);

  const serial = String((count || 0) + 1).padStart(3, "0");
  return `${prefix}${serial}`;
}

// ─── Create User ──────────────────────────────────────────────────────────────

export async function createUser(req: Request, res: Response): Promise<void> {
  const { role, full_name, email, phone, password, region, school, subject, class_grade } =
    req.body;
  const creatorUserId = (req as any).user?.user_id;

  if (!role || !full_name || !email || !phone || !password) {
    res.status(400).json({ message: "Missing required fields" });
    return;
  }

  if (password.length < 8) {
    res.status(400).json({ message: "Password must be at least 8 characters" });
    return;
  }

  // Get creator's super_admin_id (FK used in role tables)
  const { data: creators } = await supabase
    .from("user_auth")
    .select("id")
    .eq("user_id", creatorUserId)
    .limit(1);
  const creatorUserAuthId = creators?.[0]?.id ?? null;

  const { data: superAdmins } = creatorUserAuthId
    ? await supabase.from("super_admin").select("admin_id").eq("user_id", creatorUserAuthId).limit(1)
    : { data: null };
  const creatorId = superAdmins?.[0]?.admin_id ?? null;

  // Generate user_id
  const userId = await generateUserId(role, school, class_grade);

  // Hash password
  const passwordHash = await bcrypt.hash(password, 10);

  const roleLabels: Record<string, string> = {
    subadmin: "Sub Admin",
    teacher: "Teacher",
    principal: "Principal",
    student: "Student",
  };
  const roleLabel = roleLabels[role] || role;

  // Insert into user_auth
  const { data: authData, error: authError } = await supabase
    .from("user_auth")
    .insert({
      user_id: userId,
      full_name,
      email,
      phone,
      password_hash: passwordHash,
      role: roleLabel,
      is_locked: false,
      failed_login_attempts: 0,
    })
    .select("id")
    .single();

  if (authError) {
    res.status(500).json({ message: "Failed to create user auth", detail: authError.message });
    return;
  }

  const userAuthId = authData.id;

  // Insert into role-specific table
  let roleError: string | null = null;

  if (role === "subadmin") {
    const { data: regions } = await supabase
      .from("region")
      .select("region_id")
      .ilike("name", `%${region || ""}%`)
      .limit(1);
    const regionId = regions?.[0]?.region_id ?? null;

    const { error } = await supabase.from("sub_admin").upsert({
      user_id: userAuthId,
      region_id: regionId,
      created_by_super_admin_id: creatorId,
      full_name,
      email,
      phone,
      is_active: true,
    }, { onConflict: "email" });
    if (error) roleError = error.message;

  } else if (role === "teacher") {
    const { data: schools } = await supabase
      .from("school")
      .select("school_id")
      .ilike("school_name", `%${school || ""}%`)
      .limit(1);
    const schoolId = schools?.[0]?.school_id ?? null;

    const { error } = await supabase.from("teacher").upsert({
      user_id: userAuthId,
      school_id: schoolId,
      full_name,
      subject_expertise: subject || null,
      is_active: true,
    }, { onConflict: "user_id" });
    if (error) roleError = error.message;

  } else if (role === "principal") {
    const { data: schools } = await supabase
      .from("school")
      .select("school_id")
      .ilike("school_name", `%${school || ""}%`)
      .limit(1);
    const schoolId = schools?.[0]?.school_id ?? null;

    const { error } = await supabase.from("principal").upsert({
      user_id: userAuthId,
      school_id: schoolId,
      full_name,
      designation: "Principal",
      is_active: true,
    }, { onConflict: "user_id" });
    if (error) roleError = error.message;

  } else if (role === "student") {
    const { data: schools } = await supabase
      .from("school")
      .select("school_id")
      .ilike("school_name", `%${school || ""}%`)
      .limit(1);
    const schoolId = schools?.[0]?.school_id ?? null;

    const { error } = await supabase.from("student").upsert({
      user_id: userAuthId,
      school_id: schoolId,
      full_name,
      class: class_grade || null,
      is_active: true,
    }, { onConflict: "user_id" });
    if (error) roleError = error.message;

  } else {
    await supabase.from("user_auth").delete().eq("id", userAuthId);
    res.status(400).json({ message: "Invalid role" });
    return;
  }

  // Rollback user_auth if role insert failed
  if (roleError) {
    await supabase.from("user_auth").delete().eq("id", userAuthId);
    res.status(500).json({ message: `Failed to create ${roleLabel}`, detail: roleError });
    return;
  }

  res.status(201).json({
    message: "User created successfully",
    user_id: userId,
    full_name,
    role: roleLabel,
  });
}

File: backend authentication/controllers/user_controller.py
import re
import bcrypt as _bcrypt
from fastapi import HTTPException
from config.supabase_client import supabase


def _hash(plain: str) -> str:
    return _bcrypt.hashpw(plain.encode(), _bcrypt.gensalt(10)).decode()


# ─── User ID Generator ─────────────────────────────────────────────────────────
async def generate_user_id(role: str, school: str = None, class_grade: str = None) -> str:
    if role == "subadmin":
        prefix = "SA"
    elif role == "teacher":
        prefix = "TA"
    elif role == "principal":
        letter = (school or "X").strip()[0].upper()
        prefix = f"P{letter}"
    elif role == "student":
        letter = (school or "X").strip()[0].upper()
        clean_class = re.sub(r"class\s*", "", class_grade or "", flags=re.IGNORECASE)
        clean_class = re.sub(r"[\s\-]", "", clean_class).upper()
        prefix = f"S{letter}{clean_class}"
    else:
        prefix = "USR"

    response = supabase.from_("user_auth").select("*", count="exact").like("user_id", f"{prefix}%").execute()
    count = response.count or 0
    serial = str(count + 1).zfill(3)
    return f"{prefix}{serial}"


# ─── Create User ───────────────────────────────────────────────────────────────
async def create_user(body: dict, current_user: dict):
    role = body.get("role")
    full_name = body.get("full_name")
    email = body.get("email")
    phone = body.get("phone")
    password = body.get("password")
    school = body.get("school")
    class_grade = body.get("class_grade")

    if not role or not full_name or not email or not phone or not password:
        raise HTTPException(status_code=400, detail="Missing required fields")

    if len(password) < 8:
        raise HTTPException(status_code=400, detail="Password must be at least 8 characters")

    role_labels = {
        "subadmin": "Sub Admin",
        "teacher": "Teacher",
        "principal": "Principal",
        "student": "Student",
    }
    role_label = role_labels.get(role, role)

    if role not in role_labels:
        raise HTTPException(status_code=400, detail="Invalid role")

    user_id = await generate_user_id(role, school, class_grade)
    password_hash = _hash(password)

    auth_res = supabase.from_("user_auth").insert({
        "user_id": user_id,
        "full_name": full_name,
        "email": email,
        "phone": phone,
        "password_hash": password_hash,
        "role": role_label,
        "is_locked": False,
        "failed_login_attempts": 0,
    }).execute()

    if not auth_res.data:
        raise HTTPException(status_code=500, detail="Failed to create user")

    return {
        "message": "User created successfully",
        "user_id": user_id,
        "full_name": full_name,
        "role": role_label,
    }
